src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -0,0 +1,518 @@
+            # If we're here, it's a delete, and the read didn't get an
+            # error, or it's a creation and we did get an error, so we
+            # need to wait and then retry
+            await asyncio.sleep(interval)
+            elapsed += interval
+        raise asyncio.TimeoutError(
+            f"Timeout while waiting for {item} {operation}"
+        )
+
+    async def copy_pull_secret(self, source_ns: str) -> Dict[str, str]:
+        secret = await self.read_secret(
+            name="pull-secret", namespace=source_ns
+        )
+        return secret.data
+
+    async def merge_controller_secrets(
+        self, secret_list: List[LabSecret], token: str, source_ns: str
+    ) -> Dict[str, str]:
+        """Merge the user token with whatever secrets we're injecting
+        from the lab controller environment."""
+        secret_names: List[str] = list()
+        secret_keys: List[str] = list()
+        for sec in secret_list:
+            secret_names.append(sec.secret_name)
+            if sec.secret_key in secret_keys:
+                raise RuntimeError("Duplicate secret key {sec.secret_key}")
+            secret_keys.append(sec.secret_key)
+        # In theory, we should parallelize the secret reads.  But in practice
+        # it makes life a lot more complex, and we probably just have one,
+        # the controller secret.  Pull-secret will be handled separately.
+        base64_data: Dict[str, str] = dict()
+        for name in secret_names:
+            secret: Secret = await self.read_secret(
+                name=name, namespace=source_ns
+            )
+            # Retrieve matching keys
+            for key in secret.data:
+                if key in secret_keys:
+                    base64_data[key] = secret.data[key]
+        # There's no point in decoding it; all we're gonna do is pass it
+        # down to create a secret as base64 anyway.
+        if "token" in base64_data:
+            raise RuntimeError("'token' must come from the user token")
+        if not token:
+            raise RuntimeError("User token cannot be empty")
+        base64_data["token"] = str(base64.b64encode(token.encode("utf-8")))
+        return base64_data
+
+    async def create_secret(
+        self,
+        name: str,
+        namespace: str,
+        data: Dict[str, str],
+        secret_type: str = "Opaque",
+        immutable: bool = True,
+    ) -> None:
+        secret = V1Secret(
+            data=data,
+            type=secret_type,
+            immutable=immutable,
+            metadata=self.get_std_metadata(name),
+        )
+        await self.api.create_namespaced_secret(namespace, secret)
+        return
+
+    async def read_secret(
+        self,
+        name: str,
+        namespace: str,
+    ) -> Secret:
+        secret: V1Secret = await self.api.read_namespaced_secret(
+            name, namespace
+        )
+        secret_type = secret.type
+        return Secret(data=secret.data, secret_type=secret_type)
```

Comment:
What happens if the secret doesn't exist? I think right now it throws an uncaught exception back up to the service layer, but we probably want to log something more explicit since this will be a not-uncommon configuration error.

Reply:
Logged.